### ISSUE TITLE
Is it possible to use this algorithm for input data combination

### ISSUE THREAD

**Issue by leolin65** [rickgroen/cov-weighting#8]
Hi Rick,

Thanks for your algorithm to let me think more.
I want to prepare 10 pictures to input to model per batch.
Basliclly, I always generate these 10 pictures from 4 argumentaion methods randomly.
for example:  " 10 images = 2  images  * Method1 + 3  images  *Method2 +3  images *Method3 +2  images *Method4 " above values(a=2, b=3, c=3, d=2)  are generated randomly. 
Is it possible to leverage your algorithm to  get these learnable values  [a,b,c,d] 
 to get best accuracy of combination ?
thank you very much.

**Comment by rickgroen** [comment]
Hi Leolin,

I don't fully understand your question, so let me try to rephrase: You have 4 augmentation methods, which in total should generate N images. The augmented images form your training set (restricted to only be N images), and the goal is to have a distribution of N images that optimizes some other metric on a validation set? 
So to summarize: Your goal is to optimize for augmentation rather than loss weighting?

This is a use-case that wasn't tested in the paper, since the paper is about finding loss weighting. 
Possibly you could formulate your problem as such? For example: does each augment attempt to guide learning in a direction that is strongly correlated with a specific loss?

Maybe it is even more straight forward to record loss statistics for each augmentation group, and assign a weight to those groups using CoVWeighting. This is possible, but does require you to re-write a bit of code. In that case, you need to keep track of which image was generated using which augmentation method. You can still update the weights at every iteration.

Rick

**Comment by leolin65** [comment]
Hi Rick,

Thank you for your kindly support.
I think some days how to modify it
Below is my code framework, how could I use CoVWeighting to embeded it?
thank you very much.

==========================================================
import torch
from torch.optim import SGD

# Define the number of images to generate
num_images = N

# Define the optimizer
optimizer = SGD(model.parameters(), lr=0.01, momentum=0.9)

# Define the learning rate scheduler
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=num_images)

# Define the loss function
loss_covweighting_class = CoVWeightingLoss(args)
    
# Train the model
for i in range(num_images):
    # Get the current learning rate
    lr = scheduler.get_lr()[0]

    # Perform the augmentation
    augmented_images = perform_augmentation(images, lr)

    # Forward pass
    preds = model(augmented_images)
    multi_loss = loss_covweighting_class(preds, targets)
    
# Backward pass and optimization
    optimizer.zero_grad()
    multi_loss.backward()
    optimizer.step()

    # Update the learning rate
    scheduler.step()

**Comment by rickgroen** [comment]
I can give a couple of tips, but this is a broad question.
1) Your augment_images should pass which augmentation technique is used for which image.
2) You should modify your BaseLoss (super class of CoVWeighting) to compute losses per augment group, rather than per different loss. If you use multiple losses, you need to think about aggregation in some form.
3) You should modify your CovWeighting class to update statistics per augment class. Remember that this is not necessarily balanced per batch, so you should normalize. If you still have multiple different losses, you could also record statistics per augment class per loss. This would complicate matters, so I wouldn't start there.

As a last comment: part of research is of course to work out these things yourself. You are modifying the code beyond its intended use, so I cannot offer any support beyond general tips.
With that I think I want to close this issue. If, in the future, you have specific questions that fall within the scope of the code, feel free to ask.
